proposals/0293-extend-property-wrappers-to-function-and-closure-parameters.md

```diff
@@ -16,193 +48,278 @@ Property Wrappers were [introduced in Swift 5.1](https://github.com/apple/swift-
 
-### Memberwise initialization
+Property wrappers attached to parameters have a wide variety of use cases. We present a few examples here.
 
-Currently, property-wrapper attributes on struct properties interact with function parameters through the struct's synthesized memberwise initializer. However, property-wrapper attributes are _not_ supported on function parameters. This leads to complicated and nuanced rules for which type, between the wrapped-value type and the backing property-wrapper type, the memberwise initializer accepts.
+### Argument validation
 
-The compiler will choose the wrapped-value type to offer a convenience to the call-site when the property wrapper has an initializer of the form `init(wrappedValue:)` accepting the wrapped-value type, as seen here:
+Both library developers and language users often need to assert their assumptions, for which `precondition(_:_:)` is often used:
 
 ```swift
-import SwiftUI
-
-
-struct TextEditor {
-
-  @State var document: Optional<URL>
-  
+enum Product {
+  case plainSandwich 
+  case grilledCheeseSandwich 
+  case avocadoToast
 }
 
-
-func openEditor(with swiftFile: URL) -> TextEditor {
-  TextEditor(document: swiftFile) 
-  // The wrapped type is accepted here.
+func buy(quantity: Int, of product: Product) {
+  precondition(quanity >= 1, "Invalid product quanity.")
+  
+  if quantity == 1 {
+    ...
+  }
 }
 ```
 
-However, this can take flexibility away from the call-site if the property wrapper has other `init` overloads, because the call-site _cannot_ choose a different initializer. Further, if the property wrapper is explicitly initialized via `init()`, then the memberwise initializer will choose the backing-wrapper type, even if the wrapper supports `init(wrappedValue:)`. This results in unnecessary boilerplate at call-sites that _do_ want to use `init(wrappedValue:)`:
+The above code is quite clear; it has, though, the obvious drawback that changing the the condition to be asserted or its error message requires significant effort as a precondition statement is individually written for each function and manually documented.
+
+Furthermore, supposing the above is library code, we may want to test for our precondition while offering an easy-to-debug way. So, using `Validation` from [`ValidatedPropertyKit`](https://github.com/SvenTiigi/ValidatedPropertyKit) we can write:
 
 ```swift
-import SwiftUI
+@propertyWrapper
```

Comment:
ValidatedPropertyKit already defines a `@Validated` wrapper, hence my naming `@Asserted`. I tried to avoid reusing that name so as to avoid confusion. I think 'Asserted' could definitely be improved. Do you have any other ideas for a name?

Reply:
Oh I see. I was confused because the function below this used `Validated` so I thought it was just a mistake, sorry! I do think it would be fine to use (a variation of) the `Validated` property wrapper from the package, but please feel free to change the name to whatever you think is best